### ISSUE TITLE
IE support: Array.find is missing

### ISSUE THREAD

**Issue by skyflyer** [qodesmith/datepicker#48]
With reference to #19, #43, we've found out that Array.find is not supported. [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/find) includes a simple polyfill, which could be used, or the code could be adapted to use `.indexOf` instead.

The code that uses `.find`: https://github.com/qodesmith/datepicker/search?q=.find&unscoped_q=.find

I can prepare a PR if you'd like.

**Comment by qodesmith** [comment]
Looks like I have to tweak my Babel setup. I'll have a look & push a fix. Also, if there's any other IE related issues, please let me know! Thanks.

**Comment by qodesmith** [comment]
Fixed. Instead of going down the polyfill route with Babel & friends, I simply switched to other supported array methods to accomplish the same tasks. You can `npm install` to the latest version, 4.0.8. Let me know if this fixes the issue. I'll keep this open for a while.

**Comment by skyflyer** [comment]
So I played a little with weback on windows and IE11. After changing fat arrow syntax in sandbox.js and introducing `@babel/polyfill` datepicker started working in IE11. I don't anything about `@babel/polyfill` but it is apparently required if you want to support older browsers. I added `@babel/polyfill` into the `entry` array, as documented on their page: https://babeljs.io/docs/en/babel-polyfill

Given that you use `present-env`, I also tried adding `ie > 11` to the `browserlist` (package.json) and the `useBuiltIns: 'entry'` to the `preset-env` options; the babel reported that `import '@babel/polyfill' was not found.`, so I suspect that I misread that part of the docs;

In a nutshell, after playing with this for quite some time, I believe the "correct" way would be to:

1. `npm i -D @babel/polyfill` - it needs to be installed
1. `Add `'@babel/polyfill'` to `entry` arrray (webpack config)
1. Include `not dead` in `browserlist` (package.json)
1. set `useBuiltIns` to `false` or not set it at all (it default to false) -- it is currenlty not set, so step not needed :)

I'm interested to hear your thoughts about this.

**Comment by qodesmith** [comment]
@skyflyer If you're referring to adding all this to your own project, then yes, I think this is the right way. But this should have nothing to do with getting datepicker.js to work correctly. My concern is for it to work out of the box, so after the push I made today any errors you're seeing should be a result of other things in your project. A few thoughts though...

I'm not familiar with the `not dead` setting for `browserlist` field in package.json.  For your case I'd do something either like `last 2 versions` or specifically `ie > 9. Some good insight on why you might *not* want to use `last 2 versions` can be found [here](https://goo.gl/2uAdKL), but that mostly revolves around not having to support IE.

As for the rest, I was originally going down this road this morning for datepicker.js, but it was much easier to implement my solution. However, here are the things I did to get it to work (which are very close to your findings):

1. `npm i -D @babel/polyfill`
2. Add `import @babel/polyfill` at the top of your entry file (whatever file that is)
3. Added `last 2 versions` to the `browserlist` field in package.json
4. Used this setup for `present-env` (which is in my webpack config file):

```javascript
module: {
  rules: [
    {
      test: /\.js$/,
      include: path.resolve(__dirname, 'src'),
      use: {
        loader: 'babel-loader',
        options: {
          presets: [
            [
              '@babel/preset-env',
              {
                useBuiltIns: 'entry'
              }
            ]
          ]
        }
      }
    }
  ]
}
```

I think what is happening is that with this webpack setup, babel will see the import statement at the top of your entry file (either `import '@babel/polyfill'` or `require('@babel/polyfill')`) and *replaces* it with the necessary polyfills. Actually, I think the `'entry'` value causes babel to include *every* polyfill it has, but if you use `'usage'` it will automatically detect what you're using and polyfill that.

The reason I didn't go this route was because I diffed the old minified file with the new one and it added a tooon to the file. So in my case, it was much better to do what I did. But the babel docs even tell you not to include the polyfill wholesale if you're developing a library or tool. It's really meant for projects, like what you're working on. Hope that helps!

**Comment by skyflyer** [comment]
@qodesmith, I guess my comment was not very clear and perhaps I even misunderstood the intent of the library.

The js-datepicker, as it is with version `4.0.8` does not work in IE 11, when referenced. I'm not sure whether you were able to test it in IE11 tested that or not, but the error reported is `'Symbol' is undefined` and I believe this is due to babel transpiling.

It is my understanding that this library is supposed to be used directly in the browser as well as integrated - and this case is currently failing for IE11 users.

I created a simple HTML page and referenced CSS+JS from unpkg directly, version 4.0.8. IE barfs.

----

I usually don't work with babel, so I don't have a good insight into the tooling, etc; and I've read the same comment about library authors, although after my yesterday's research, I'm inclined to believe that comment is geared toward libraries which are not used in-browser.

With regards to `last 2 versions`: I had assumed that IE11 fits this criteria, as it is actually the last IE version, but my testing showed otherwise. I've gone and re-read the browserlist docs and tried again with only `last 2 versions` and now IE works -- I suppose my babel/polyfill was not configured properly at the time, as I was changing more than one thing at a time. So sorry about confusion regarding this one.

What are your thoughts on the first section?

**Comment by qodesmith** [comment]
Now I see what you mean. Yes, you are absolutely correct. This library should be able to be used right out of the box in IE 11. I'll look into it.

**Comment by qodesmith** [comment]
So I was a bit confused why Babel was even putting the Symbol stuff in there. Turns out that if you use the spread operator in your code - `{ ...obj1, obj2 }` - then Babel uses the Symbol stuff. I'm going to take another pass at my source code and see if I can refactor a bit. I want to get to the point where Babel is only transpiling ES6 => ES5 without the need for any polyfills. Stay tuned...

**Comment by skyflyer** [comment]
>  I want to get to the point where Babel is only transpiling ES6 => ES5 without the need for any polyfills. 

This makes perfect sense. Out of curiosity: why aren't you using Typescript? To me, that is pure awesomeness :) No babel required.

**Comment by qodesmith** [comment]
Eh, I'm not a fan of the TS syntax 😛 Try the latest version, `4.0.9` and let me know if that solved the issue. Fingers crossed...

**Comment by qodesmith** [comment]
It's still not working. I have IE11 on my home computer but I have to leave now. I'll fool around with Babel settings and see if that works.

**Comment by skyflyer** [comment]
Just tested, the `Symbol` is still present. :(

**Comment by qodesmith** [comment]
Ok, I just removed `@babel/preset-env` since that's mainly to allow you to use the latest JS and I'm not really doing anything fancy but some basic ES6 syntax. Since I'm not at home and can't test on IE11, I looked at the minified code and this ended up removing a reference to Symbol. Try it again and let me know how things go.

**Comment by qodesmith** [comment]
Turns out, after some deep searching, it's array destructuring that causes Babel to use Symbol. I'm looking into how to *only* include that polyfill as I don't want to bloat this package.

**Comment by skyflyer** [comment]
Hello Aaron! Did you manage to get this working in some way with babel? I understand that you don't want to drag the polyfill baggage into the final package; not sure if there is a way around it, unless not using array destructuring syntax?

**Comment by qodesmith** [comment]
Hey, sorry about the delay. Been a bit busy here. Prepping for a talk at the ReactNYC meetup tomorrow. I'm dusting off the mic and spitting 48 bars of fuego with the beat done by Ken Wheeler. Anyways, yea, I might have to get rid of array destructuring. But my instinct is there's gotta be a way to include *just* that transformation with Babel instead of all that other stuff.  I'll pick it back up on Wednesday.

**Comment by skyflyer** [comment]
Hey Aaron! I hope your talk went well. Will you be able to find some time to fix this? Is there anything I can do to help?

**Comment by skyflyer** [comment]
Aaron, happy New Year! :)

Will you be able to test this, or should I create my own private release with the babel fix?

**Comment by litemotiv** [comment]
I'm running into the same issue where 4.0.10 is breaking on IE11, using the unpkg version:

https://unpkg.com/js-datepicker@4.0.10/dist/datepicker.min.js

IE11 produces a syntax error on the fat arrow on column 1583.

**Comment by litemotiv** [comment]
@skyflyer, i'm using the minified datepicker standalone as an include so not through npm

**Comment by qodesmith** [comment]
@skyflyer @litemotiv I haven't been able to prioritize time for this. If you have a fix outside the including this library in your own bundle, feel free to make a PR. Otherwise, I'll take another swing when things calm down over here.

**Comment by qodesmith** [comment]
Just settled down in a new job so I'll be taking another look at this. Going to use Browserstack to emulate older versions of IE to iron out the kinks.

**Comment by qodesmith** [comment]
Just updated the library. Let me know if you continue to have any issues.

**Comment by behler-young-insite** [comment]
I had the same problem but after importing core-js on my index.ts I didn't see the Symbol error anymore - though the bloat factor did jump a bit.

**Comment by epxfrho** [comment]
Hi, I still get the "symbol is undefined" error in ie11, using the minified datepicker version 5.2.1. 

I've also tested it stand-alone outside my application:

```
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>js-datepicker-test</title>
    <link rel="stylesheet" href="datepicker.min.css">
    <script src="datepicker.min.js"></script>
  </head>
  <body>
    <h1>js-datepicker-test</h1>
    <span id="testdatepicker">datepicker</span>
    <script>
      var test = datepicker('#testdatepicker');
    </script>
  </body>
</html>
```

error: **symbol is undefined** nearby '... function(e,t){if(!(Symbol.iterator in Object(e) ...' in datepicker.min.js
